### PR TITLE
Don't Collect UDQ Values if None Exist

### PR DESCRIPTION
**Pull request by bska**
This PR switches the `AggregateUDQData::dUDx_` members to being `std::optional<>` instead of direct `WindowedArray<>` objects.  This enables using `std::nullopt` to represent "no such UDQ category exists" and makes the output logic in `RestartIO.cpp:writeUDQ()` slightly less magic.

While here, also reorder the `AggregateUDQData.cpp` implementation file so that the dUdx namespaces appear in alphabetical order.

Finally, move the `ig_phase()` helper function to the .cpp file since it does not need to be public and switch the `AggregateUDQData` constructor to taking a `UDQDims` object directly instead of `UDQDims.data()`.  This enables removing the mostly trivial `createUDQDims()` helper function, albeit at a (temporary) cost of reduced readability in expressions that expected the vector semantics of its return value.

This work is in preparation of adding support for segment level UDQs in the restart file.

### CODE REVIEW

**Comment by bska** [comment]
I'm creating the PR in draft mode because it depends on, and contains, the earlier PR #4134. I will keep this PR in a draft state until such time as it is ready for review.

**Comment by bska** [comment]
jenkins build this please

File: opm/output/eclipse/AggregateUDQData.cpp [hunk 1]
```diff
@@ -572,356 +749,379 @@ namespace {
         allocate(const std::vector<int>& udqDims)
         {
             using WV = Opm::RestartIO::Helpers::WindowedArray<int>;
-            int nwin = std::max(udqDims[7], 1);
+
+            const int nwin = std::max(udqDims[7], 1);
```
Comment (akva2) [inline opm/output/eclipse/AggregateUDQData.cpp:753]: I find this expression kinda confusing, is udqDims[7] >= 0 or not ?  if it is i'd find a ternary more expressive, if not the return value below seems suspicious.

**Review by akva2 — APPROVED** [review]
Some questions/suggestions, since lines are touched anyways.

File: opm/output/eclipse/AggregateUDQData.cpp [hunk 2]
```diff
@@ -405,6 +453,119 @@ namespace {
...
+            else if ((wg_key == Opm::UDAKeyword::GCONPROD) ||
+                     (wg_key == Opm::UDAKeyword::GCONINJE))
+            {
+                if (iuap[ind].wgname != "FIELD") {
+                    const auto& group = sched.getGroup(iuap[ind].wgname, simStep);
+                    wg_no.push_back(group.insert_index() - 1);
+                }
+            }
+            else {
+                std::cout << "Invalid Control keyword: "
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/output/eclipse/AggregateUDQData.cpp]: I've pushed an update to use `OpmLog` instead.

File: opm/output/eclipse/AggregateUDQData.cpp [hunk 3]
```diff
@@ -209,18 +247,19 @@ namespace {
         while (search_pos < modTokens.size()) {
             if (modTokens[search_pos].type() == Opm::UDQTokenType::open_paren  && level == 0) {
                 startParen.emplace_back(search_pos);
-                level +=1;
+                level += 1;
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/output/eclipse/AggregateUDQData.cpp]: I've pushed an update to switch to pre-increment instead.

File: opm/output/eclipse/AggregateUDQData.cpp [hunk 4]
```diff
@@ -405,6 +453,119 @@ namespace {
...
     }
 
+    std::vector<int>
+    ig_phase(const Opm::Schedule&    sched,
+             const std::size_t       simStep,
+             const std::vector<int>& inteHead)
+    {
+        std::vector<int> inj_phase(ngmaxz(inteHead), 0);
+
+        auto update_phase = [](int& phase, int new_phase) {
```
[the remark this answers is not in the document]
Comment (bska) [reply opm/output/eclipse/AggregateUDQData.cpp]: I've pushed an update to remove the out-parameter here, as well as reversing some of the checks to reduce the level of nesting.

File: opm/output/eclipse/AggregateUDQData.cpp (hunk 1 above)
Comment (bska) [reply opm/output/eclipse/AggregateUDQData.cpp]: The original code was somewhat questionable and pushed the [check for non-zero size]() to `RestartIO.cpp`.  I've pushed an update that makes the UDA-related arrays `optional<>` instead which means we get the same kind of call site logic for those as we do for the `DUDx` arrays.

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
jenkins build this please

**Comment by bska** [comment]
PR approved and build check is green.  I'll merge into master.

**Pull request merged**